pkg/app/api/grpcapi/web_api.go

```diff
@@ -168,6 +170,107 @@ func (a *WebAPI) ListEnvironments(ctx context.Context, req *webservice.ListEnvir
+		return nil, err
+	}
+	return &webservice.DisableEnvironmentResponse{}, nil
+}
+
+func (a *WebAPI) DeleteEnvironment(ctx context.Context, req *webservice.DeleteEnvironmentRequest) (*webservice.DeleteEnvironmentResponse, error) {
+	claims, err := rpcauth.ExtractClaims(ctx)
+	if err != nil {
+		a.logger.Error("failed to authenticate the current user", zap.Error(err))
+		return nil, err
+	}
+
+	if err := a.validateEnvBelongsToProject(ctx, req.EnvironmentId, claims.Role.ProjectId); err != nil {
+		return nil, err
+	}
+
+	if err := a.environmentStore.DeleteEnvironment(ctx, req.EnvironmentId); err != nil {
+		switch err {
+		case datastore.ErrNotFound:
+			return nil, status.Error(codes.NotFound, "The environment is not found")
+		case datastore.ErrInvalidArgument:
+			return nil, status.Error(codes.InvalidArgument, "Invalid value to delete")
+		default:
+			a.logger.Error("failed to delete the environment",
+				zap.String("env-id", req.EnvironmentId),
+				zap.Error(err),
+			)
+			return nil, status.Error(codes.Internal, "Failed to delete the environment")
+		}
+	}
+
+	return &webservice.DeleteEnvironmentResponse{}, nil
+}
+
+func (a *WebAPI) updateEnvironmentEnable(ctx context.Context, envID string, enable bool) error {
+	claims, err := rpcauth.ExtractClaims(ctx)
+	if err != nil {
+		a.logger.Error("failed to authenticate the current user", zap.Error(err))
+		return err
+	}
+
+	if err := a.validateEnvBelongsToProject(ctx, envID, claims.Role.ProjectId); err != nil {
+		return err
+	}
+
+	var updater func(context.Context, string) error
+	if enable {
+		updater = a.environmentStore.EnableEnvironment
+	} else {
+		updater = a.environmentStore.DisableEnvironment
+	}
+
+	if err := updater(ctx, envID); err != nil {
+		switch err {
+		case datastore.ErrNotFound:
+			return status.Error(codes.InvalidArgument, "The environment is not found")
+		case datastore.ErrInvalidArgument:
+			return status.Error(codes.InvalidArgument, "Invalid value for update")
+		default:
+			a.logger.Error("failed to update the environment",
+				zap.String("env-id", envID),
+				zap.Error(err),
+			)
+			return status.Error(codes.Internal, "Failed to update the environment")
+		}
+	}
+	return nil
+}
+
+// validateEnvBelongsToProject checks if the given piped belongs to the given project.
+// It gives back error unless the env belongs to the project.
+func (a *WebAPI) validateEnvBelongsToProject(ctx context.Context, envID, projectID string) error {
+	eid, err := a.envProjectCache.Get(envID)
+	if err == nil {
+		if eid != projectID {
+			return status.Error(codes.PermissionDenied, "Requested environment doesn't belong to the project you logged in")
+		}
+		return nil
+	}
+
+	env, err := getEnvironment(ctx, a.environmentStore, envID, a.logger)
+	if err != nil {
+		return err
+	}
+	a.envProjectCache.Put(envID, env.ProjectId)
+
+	if env.ProjectId != projectID {
+		return status.Error(codes.PermissionDenied, "Requested environment doesn't belong to the project you logged in")
+	}
```

Comment:
should we validate by L268 if condition before update the cache 🤔

Reply:
I would say no. `envID` and `env.ProjectId` at L266 are exactly the correct pair.
The validation target at this function is `projectID`. Why you got confused may be the If condition order; we should make the order as:
```go
if projectID != env.ProjectId
```

Reply:
oh, I get your point now 👍  Thanks for clarifying 🙏